source/_posts/2016-07-18-six-useful-ember-addons-for-testing.markdown

```diff
@@ -0,0 +1,224 @@
+
+Acceptance tests are great because with little code they allow you to test a big part of your application almost end-to-end (at least at client-side level). One problem though is that they can become unmaintainable really fast if you don't give the tests a proper structure.
+
+[ember-cli-page-object](http://ember-cli-page-object.js.org/) aids you by providing a set of well known practices and conventions that help you avoid common mistakes like tying your tests to the HTML structure of your application and by avoiding the repetition of lots of code (CSS selectors and DOM manipulation helpers).
+
+Here's a glimpse of how you can write an acceptance test using a page object.
+
+```js
+const page = PageObject.create({
+  visit: visitable('/'),
+
+  username: fillable('#username'),
+  password: fillable('#password'),
+  submit: clickable('button'),
+  error: text('.errors')
+});
+
+test('login failed', function(assert) {
+  page
+    .visit()
+    .username('admin')
+    .password('invalid')
+    .submit();
+
+  andThen(() => {
+    assert.equal(page.error, 'Invalid credentials');
+  });
+});
+```
+
+This addon also supports component integration tests so you can use the same abstractions that you use in your acceptance tests.
+
+## <a name="ember-try"></a>ember-try
+
+Included by default in Ember addon projects, this awesome addon by [Katie Gengler](https://twitter.com/katiegengler) (of emberobserver.com fame!) allows you to run your test suite against multiple versions of a dependency.
+
+Let's say that you create an addon and want to ensure that it won't break with future releases of Ember, you can declare in ember-try's configuration file that you would like to run the tests using Ember Canary, Beta, Release and why not, Ember LTS. The addon adds a new command to ember-cli which reads said configuration file and runs your entire test suite several times but changing the ember version in each run to cover all the different versions you specified. This way you can ensure that your addon will continue working when newer versions of Ember are released.
+
+Apart from testing against different versions of Ember, you can also test against different versions of other libraries, let's say you want to ensure that your addon runs correctly using different versions of jQuery, you can do this really easily also.
+
+You can find more information about the addon [here](https://github.com/ember-cli/ember-try).
+
+## <a name="ember-exam"></a>ember-exam
+
+[ember-exam](https://github.com/trentmwillis/ember-exam) is a much needed and awaited project by [Trent Willis](https://twitter.com/trentmwillis), a QUnit core developer. The addon adds a new ember-cli command which allows you to run your tests with randomization, splitting, and parallelization which reduces the time to run your complete test suite and also helps you spot flaky tests.
+
+After you install the addon you just use the command `exam`
+
+```
+$ ember exam --random --split=4
```

Comment:
Should be `ember exam --random --split=4 --parallel` to run the tests in parallel